### PR TITLE
fixes #3011 - default root node name for api index json response to :results but make it configurable

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by domcleal** [comment]
There's another couple here:

<pre>
app/views/api/v2/compute_resources/available_images.json.rabl
1:collection @available_images, :object_root => :image
app/views/api/v2/import_puppetclasses/index.json.rabl
1:collection @environments, :root => false, :object_root => false
</pre>

Also the API show responses [should also be 'result'](http://theforeman.org/manuals/1.3/index.html#5.1.4JSONResponseforSingleObjects), are you addressing this separately?  It seems fairly related.

**Comment by isratrade** [comment]
I'm not set up having 'result' for single objects.  I prefer either the name of the object or nothing at all.  I also plan to soon go through and update the API v2 docs on theforeman.org

**Comment by domcleal** [comment]
The decision here seemed to be for a configurable 'result' node:
https://github.com/theforeman/theforeman.org/pull/106#discussion_r7252476

**Comment by isratrade** [comment]
@domcleal, in any case, the single object response with be in a separate PR.  This PR is only on collections and is what is required for katello nutupane.

**Comment by isratrade** [comment]
@domcleal, I updated available_images.json.rabl but left as is, since it doesn't use the index_layout.json.rb
app/controllers/concerns/api/import_puppetclasses_common_controller.rb#32

``` ruby
render("api/v1/import_puppetclasses/#{rabl_template}", :layout => "api/layouts/import_puppetclasses_layout")
```

**Comment by domcleal** [comment]
Okay, filed the single object response as [#3646](http://projects.theforeman.org/issues/3646).

**Comment by domcleal** [comment]
1. Could we include this change to the puppet classes hash response here too please?  https://github.com/isratrade/foreman/commit/ed7876bae5e745f37bd29bc207d66016d303b2c7
2. fact_values also needs something similar as it returns a hash with no results node
3. ditto for tasks
4. ditto for statistics
5. ditto for dashboard
6. ditto for host/hostgroup_classes I think?

**Comment by isratrade** [comment]
@domcleal 
I updated puppetclasses and fact_values to use layout.json.rabl as they are AR objects

I don't think it makes sense to to show metadata (total, subtotal, order, etc) on non-AR objects.  
tasks - left alone
statistics - removed the node :statistics
dashboard - left alone - no root node
host/:id/puppetclass_ids (hostgroup_classes) - left alone - I feel an array of ids is the correct response.

**Comment by isratrade** [comment]
@domcleal separate commit for your review and then to squash

File: app/controllers/api/v2/statistics_controller.rb [hunk 1]
```diff
@@ -18,10 +18,10 @@ def index
         @swap_free   = FactValue.my_facts.mem_average "swapfree"
         @mem_totsize = FactValue.my_facts.mem_sum "memorysize"
         @mem_totfree = FactValue.my_facts.mem_sum "memoryfree"
-        render :json => { :statistics => { :os_count    => @os_count, :arch_count => @arch_count,
+        render :json => { :os_count    => @os_count, :arch_count => @arch_count,
```
Comment (domcleal) [inline app/controllers/api/v2/statistics_controller.rb:21]: We don't have a root node on this response?

**Comment by domcleal** [comment]
There are a couple of test failures (puppetclasses, stats), plus my comment above.  otherwise :+1:

File: app/controllers/api/v2/statistics_controller.rb (hunk 1 above)
Comment (isratrade) [inline app/controllers/api/v2/statistics_controller.rb:21]: @domcleal, I removed this as I don't think that non-AR objects need to show metadata (total, subtotal, order, etc).

Same for dashboard, host/:id/puppetclass_ids -- no root node.

**Comment by isratrade** [comment]
@domcleal, I squashed the 3 commits in this PR into one and included it in PR #952

File: app/controllers/api/v2/statistics_controller.rb (hunk 1 above)
Comment (domcleal) [inline app/controllers/api/v2/statistics_controller.rb:21]: I don't really agree with this - although there's no metadata at the moment, it's inconsistent and the user then has to be aware of what responses might be AR based and which might not.

Also not documented... http://theforeman.org/manuals/1.3/index.html#5.1.2JSONResponseFormatforCollections

Comment (isratrade) [inline app/controllers/api/v2/statistics_controller.rb:21]: @domcleal, no problem.  I will put the root back.  Should it be :results or :statistics?

Comment (domcleal) [inline app/controllers/api/v2/statistics_controller.rb:21]: :results to match the others

File: app/controllers/api/v2/statistics_controller.rb [hunk 2]
```diff
@@ -18,7 +18,7 @@ def index
         @swap_free   = FactValue.my_facts.mem_average "swapfree"
         @mem_totsize = FactValue.my_facts.mem_sum "memorysize"
         @mem_totfree = FactValue.my_facts.mem_sum "memoryfree"
-        render :json => { :statistics => { :os_count    => @os_count, :arch_count => @arch_count,
+        render :json => { :results => { :os_count    => @os_count, :arch_count => @arch_count,
```
Comment (domcleal) [inline app/controllers/api/v2/statistics_controller.rb]: Ah, we might need to call root_node_name here so it's dynamic, then the user can specify the "root_name" parameter.

**Comment by domcleal** [comment]
:+1:

**Comment by domcleal** [comment]
Merged as cc739ab, thanks @isratrade!

**Pull request closed without merging**